Fuzzing strategies documentation (#815)
software-mansion/protostar: website/docs/tutorials/07-testing/03-fuzzing/strategies.md

# Strategies

> While a felt is a basis of all types in Cairo, its whole value space rarely matches what values are
> acceptable in tested code (for example, when using range checks builtin).

Reviewer: ```suggestion
Often in code we operate on some assumptions about the possible values of variables. 
For example we can perform greater than zero validation early in the code, and we can assume that variables are greater than zero in the subsequent code. 
Fuzzer allows to provide such assumptions to avoid testing against values which are not covered by the code. 
```

Reviewer: I think it better explains strategies use case

> Using the [`given`](../02-cheatcodes/given.md) cheatcode,
> one can instruct Protostar how constrained or adjusted the value space of fuzzed inputs should be.
> Such constraints are provided declaratively, by assigning _strategies_ to the input parameters.
> A strategy is an object that is capable of generating next input examples in specific way,
> and also is capable of other internal features, such as simplifying failing examples.

Reviewer: ```suggestion
Cheatcode [`given`](../02-cheatcodes/given.md) ,
instructs fuzzer how to constraint set of values code is tested against.
Such constraints are provided declaratively, by assigning _strategies_ to the input parameters as on the example below.
```

> By default, Protostar applies the [`strategy.felts()`](#strategyfelts) strategy to all felt

Reviewer: Would move this under the example and add info callout

parameters.

```cairo title="Example"
@external
func setup_integers() {
    %{
        given(
            a = strategy.integers(10, 20),
            b = strategy.integers(30, 40),
        )
    %}
    return ();
}

@external
func test_integers{syscall_ptr : felt*, range_check_ptr}(a : felt, b : felt) {
    assert_le(a, b);
    return ();
}
```

This document is a guide to what strategies are available for generating examples and how to build
them.

## Core strategies

All core strategies are contained in the [`strategy`](../02-cheatcodes/strategy.md)
cheatcode-namespace.

### `strategy.felts`

```python
def felts() -> Strategy: ...
```

Explores all possible felt values.

### `strategy.integers`

```python
def integers(
    min_value: Optional[int] = None,
    max_value: Optional[int] = None,
) -> Strategy: ...
```

> Generates integer values, possibly bounded to provided range.

Reviewer: ```suggestion
Generates integer values, possibly bounded by provided range.
```

> Assuming real numbers comparison semantics,

Reviewer: ```suggestion
Fuzzer picks integers from provided range and then converts them to felts.
```

> if `min_value` is not `None` then all values will be greater than or equal to `min_value`,

Reviewer: ```suggestion
If `min_value` is not `None` then all values will be greater than or equal to `min_value`,
```

and if `max_value` is not `None` then all values will be less than or equal to `max_value`.
When applied to field elements, the unbounded values may rarely overflow.

```python title="Examples"
strategy.integers(0, 100)
strategy.integers(max_value=3000)
strategy.integers(11)
```

## Adapting strategies

Often it is the case that a strategy does not produce exactly what is desired and a need to further
adapt the generated values arise.

> Although this could be done in tests directly, this hurts because adaptation may waste fuzzing
> cycles (by repeatedly testing same values after adaptation) and the code has to be copied in every
> test.

Reviewer: ```suggestion
```
It is useless information from user perspective

Reviewer: Discussed on slack

> There are also the [`assume`] and [`reject`] cheatcodes, but this should only be used to discard
> only few unwanted inputs.

Reviewer: ```suggestion
[`assume`] and [`reject`] cheatcodes provide simple interfaces to adapt a advanced strategy. Those are not very good considering the performance. Fuzzer can execute test on rejected data anyway and will just ignore failure if it happens.
```

Protostar provides ways to build strategies by transforming other ones.

### Mapping

```python
class Strategy:
    def map(self, mapping_function: Callable[[int], int]) -> Strategy: ...
```

Applies provided mapping function to all searched inputs values.

```python title="Example"
strategy.felts().map(lambda x: x // 2)
```

### Filtering

```python
class Strategy:
    def filter(self, filter_function: Callable[[int], bool]) -> Strategy: ...
```

Allows rejecting examples matching a condition.

> 

Reviewer: Please move example here, generally try to elevate examples higher, it's easier to understand for a reader

The outcome is similar to using the [`assume`] or [`reject`] cheatcodes, but `filter` does not
require executing tested Cairo function and thus is more performant.
Try to use `filter` only to avoid unwanted corner cases rather than attempting to cut out a large
portion of the searched input values.

> The `filter` method is not magic and hard to satisfy conditions may cause the fuzzer to fail.

Reviewer: I have no idea what this sentence means being honest

Author: `filter(lambda: 10 < x < 100)` will cause Hypothesis to raise exceptions that strategies are bad

Author: ```suggestion
Fuzzer draws random data from the original strategy and only afterwards checks if it passes filter conditions.
If too many variables are restricted, fuzzer will reject test execution.
```

Reviewer: Discussed on slack

> ```python title="Example"

Reviewer: Right now it looks like it presents
>  hard to satisfy conditions may cause the fuzzer to fail.

Author: wouldn't such example be within `warning` block?

Reviewer: I think even if we use such convention, reader does not have to be aware of it.
Also the loop of
- information
- example
- information
- example

is much beter from the cognition perspective than

-information
-information
-example
-example

It's better to keep examples close to related information

Author: ok, makes sense. thanks for tip!

# Strategies

Often in code we operate on some assumptions about the possible values of variables. 
For example we can perform greater than zero validation early in the code, and we can assume that variables are greater than zero in the subsequent code. 
Fuzzer allows to provide such assumptions to avoid testing against values which are not covered by the code. 

> Cheatcode [`given`](../02-cheatcodes/given.md) ,

Reviewer: ```suggestion
Cheatcode [`given`](../02-cheatcodes/given.md),
```
Not sure if it will render correctly

        )
    %}
    return ();
}

@external
func test_integers{syscall_ptr : felt*, range_check_ptr}(a : felt, b : felt) {
    assert_le(a, b);
    return ();
}
```

:::note
By default, Protostar applies the [`strategy.felts()`](#strategyfelts) strategy to all felt
parameters.
:::

This document is a guide to what strategies are available for generating examples and how to build
them.

## Core strategies

All core strategies are contained in the [`strategy`](../02-cheatcodes/strategy.md)
cheatcode-namespace.

### `strategy.felts`

```python
def felts() -> Strategy: ...
```

Explores all possible felt values.

### `strategy.integers`

```python
def integers(
    min_value: Optional[int] = None,
    max_value: Optional[int] = None,
) -> Strategy: ...
```

Generates integer values, possibly bounded by provided range.

```python title="Examples"
strategy.integers(0, 100)
strategy.integers(max_value=3000)
strategy.integers(11)
```

Fuzzer picks integers from provided range and then converts them to felts.
If `min_value` is not `None` then all values will be greater than or equal to `min_value`,
and if `max_value` is not `None` then all values will be less than or equal to `max_value`.

> When applied to field elements, the unbounded values may rarely overflow.

Reviewer: I am not sure what this sentence means

Author: ```suggestion
```

## Adapting strategies

> Often it is the case that a strategy does not produce exactly what is desired and a need to further
> adapt the generated values arise.

Reviewer: ```suggestion
Often it is the case that a strategy does not produce exactly what is desired and it is necessary to further
adapt the generated values.
```
Simpler to read

> Although this could be done in tests directly, this hurts because fuzzer does not know about the adaptation and may repeatedly test same values.

Author: ```suggestion
Although this could be done in tests directly, this hurts because fuzzer does not know about the adaptation and may repeatedly test the same values.
```

The [`assume`] and [`reject`] cheatcodes provide simple interfaces to adapt a advanced strategy.
Those are not very good considering the performance.
Fuzzer can execute test on rejected data anyway and will just ignore failure when it happens.

Protostar provides ways to build strategies by transforming other ones.

### Mapping

```python
class Strategy:
    def map(self, mapping_function: Callable[[int], int]) -> Strategy: ...
```

Applies provided mapping function to all searched inputs values.

```python title="Example"
strategy.felts().map(lambda x: x // 2)
```

### Filtering

```python
class Strategy:
    def filter(self, filter_function: Callable[[int], bool]) -> Strategy: ...
```

> Allows rejecting examples matching a condition.

Reviewer: ```suggestion
Rejects examples not matching a condition.
```
?

```python title="Example"
strategy.felts().filter(lambda x: x not in [3, 5, 8])
```

The outcome is similar to using the [`assume`] or [`reject`] cheatcodes, but `filter` does not
require executing tested Cairo function and thus is more performant.
Try to use `filter` only to avoid unwanted corner cases rather than attempting to cut out a large
portion of the searched input values.

Fuzzer draws random data from the original strategy and only afterwards checks if it passes filter conditions.
If too many variables are restricted, fuzzer will reject test execution.

### Combining

```python
def one_of(*strategies: Strategy) -> Strategy: ...
```

> Return a strategy which generates values from any of the argument strategies.

Reviewer: ```suggestion
Returns a strategy which generates values from any of the argument strategies.
```

```python title="Example"
strategy.one_of(
    strategy.integers(0, 100),
    strategy.integers(1000, 1200),
)

> ```

Reviewer: What happens if strategies are overlaping?
Probability for some values increases?
